### DISCUSSION TITLE
Renovate with python

### DISCUSSION THREAD

**Discussion by chamoulox** [renovatebot/renovate#12001] [Request Help]
Hello,

I would like to use renovate to automatically deals with upgrades but I'm facing some issues and I would like to know how to proceed.

Let's take a simple case:
I have a requirements file with inside main dependencies froze and transitive dependencies (also froze) created with pip freeze. It looks like:

```txt
backoff==1.10.0
requests==2.26.0
## The following requirements were added by pip freeze:
certifi==2021.5.30
chardet==4.0.0
idna==3.2
urllib3==1.26.6
```

Currently, we manually upgrade main dependencies (in these case requests  or backoff) using eager strategy.
```bash
pip install --upgrade --upgrade-strategy eager requests
```
So as a result, we upgrade transitive dependencies depending on main dependencies

With out of the box renovate, it creates one PR per dependency, so it's possible to have a PR with for example an upgrade for idna that won't be compatible with current requests version.

 I've read [this discussion](https://github.com/renovatebot/renovate/discussions/8925) that suggests creating groups for dependencies that should be upgrade together but I'm facing another issue when creating groups:
In a PR created by renovate, it doesn't take into account requirements from dependencies.
For example, if  the new version of  requests requires urlib3 <= 1.26.6 but a new version is available, it will update both in same PR, causing build issue in my pipeline.

What is the best practice to upgrade python packages with Renovate ? Is using pipenv will answer my questions ?

Thanks

**Comment by rarkins** [comment]
Can you elaborate more on the `pip freeze` step which appears to have taken place earlier?

I haven't seen this approach in pip before. Seems like it's an alternative to using pip-compile?

**Comment by chamoulox** [comment]
Yes it's similar.

Let's say I have backoff and requests as dependencies. I only install then in my virtual env, then I do `pip freeze > requirements.txt`, it will generate the file I posted, with transitive dependencies that have been added.

**Reply by rarkins** [reply]
ok gotcha! thanks. Where I was confused was I was thinking the `requirements.txt` was half generated by hand, half automated - but turns out it's fully automated from the `setup.py`. In that case, what if you simply ask Renovate to ignore the `requirements.txt` fully and instead focus on the `setup.py` - would the only problem be that the `setup.py` gets updated and the `requirements.txt` gets out of date?

**Reply by chamoulox** [reply]
Ah yes sorry, there is no setup.py, I manually add new packages then freeze dependencies.

In the case you mentioned, yes it's an issue because transitive dependencies won't be updated, so there will be some conflicts when building my app.

Maybe I should rephrase my issue:

How can I let renovate deals with direct dependencies as well as transitive dependencies; knowing that all dependencies (direct and transitive) should be frozen to avoid surprises (I had this issue before, a transitive dependency that wasn't frozen broke my app behavior)  

I saw that pip compile support is on it's way, maybe this will helps ?

**Reply by rarkins** [reply]
Pip compile would be better I think because it's more formalized. The challenge with your current setup is how Renovate can know which are transitive and which direct
